internal/impl/git/input.go

```diff
@@ -0,0 +1,709 @@
+				msg := service.NewMessage(nil)
+				relPath, err := filepath.Rel(in.tempDir, event.path)
+				if err != nil {
+					return nil, nil, fmt.Errorf("failed to get relative path for %s: %w", event.path, err)
+				}
+				msg.MetaSet("git_file_path", relPath)
+				msg.MetaSet("git_commit", in.getLastCommit().String())
+				msg.MetaSetMut("git_deleted", true)
+				return msg, func(ctx context.Context, _ error) error { event.ackFn(); return nil }, nil
+			}
+
+			msg, err := in.createMessage(event.path)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			// If createMessage returns nil, nil, it means we should skip this file
+			if msg == nil {
+				continue // Skip this file and read the next one
+			}
+
+			return msg, func(ctx context.Context, _ error) error { event.ackFn(); return nil }, nil
+		}
+	}
+}
+
+// Close implements service.Input. It signals shutdown and cleans up the temporary repository directory.
+func (in *input) Close(ctx context.Context) error {
+	if in.shutSig == nil {
+		return nil
+	}
+	in.shutSig.TriggerHardStop()
+	select {
+	case <-in.shutSig.HasStoppedChan():
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+// cloneRepo clones the configured Git repository into the temporary directory and
+// sets the initial commit hash.
+func (in *input) cloneRepo(ctx context.Context) error {
+	auth, err := in.setupAuth()
+	if err != nil {
+		return err
+	}
+
+	in.repository, err = git.PlainClone(in.tempDir, false, &git.CloneOptions{
+		URL:           in.cfg.repoURL,
+		Auth:          auth,
+		ReferenceName: plumbing.NewBranchReferenceName(in.cfg.branch),
+		SingleBranch:  true,
+		Depth:         1,
+	})
+	if err != nil {
+		return fmt.Errorf("git clone failed: %w", err)
+	}
+	ref, err := in.repository.Head()
+	if err != nil {
+		return fmt.Errorf("unable to get reference: %w", err)
+	}
+	in.lastCommitMu.Lock()
+	in.lastCommit = ref.Hash()
+	in.lastCommitMu.Unlock()
+	return nil
+}
+
+// setLastCommit sets the in.lastCommit field and updates the checkpoint cache (if configured).
+func (in *input) setLastCommit(ctx context.Context, newCommit plumbing.Hash) {
+	in.lastCommitMu.Lock()
+	in.lastCommit = newCommit
+	in.lastCommitMu.Unlock()
+
+	if in.cfg.checkpointCache == "" {
+		return
+	}
+	if err := in.updateCheckpointCache(ctx, newCommit); err != nil {
+		in.log.Errorf("failed to update checkpoint cache: %v", err)
+	}
+}
+
+// getLastCommit retrieves the lastCommit we pulled in a concurrent safe way.
+func (in *input) getLastCommit() plumbing.Hash {
+	in.lastCommitMu.RLock()
+	defer in.lastCommitMu.RUnlock()
+	return in.lastCommit
+}
+
+// pollChanges runs in a separate goroutine and periodically checks for updates
+// in the Git repository according to the configured poll interval.
+func (in *input) pollChanges(ctx context.Context, cachedCommit plumbing.Hash) {
+	hasCheckpoint := cachedCommit != plumbing.ZeroHash
+	var initialScanWg *sync.WaitGroup
```

Comment:
Nicely done with the waitgroup.
I ignored the need for acking the messages before I can mark them committed 🤦